[bug] button spacing on scratch org cards

dvdherron opened the pull request:
![](https://source.unsplash.com/featured/?cute,animal&4reel)
<!-- Swap `CHANGE_ME` for a random string to get a different image -->



## Description
Multiple buttons on Task Org cards took up too much space and caused overflow. 
Changes made in this PR make the buttons wrap when space requires it. 



## Steps to test/reproduce
- Make changes in a scratch org
- also make an org `out-of-date `
- adjust the size of the window until the parent container is 50% wide or the screen is narrow

_Please explain how to best reproduce the issue and/or test the changes locally (including the pages/URLs/views/states to review)._

## Show me
_Provide screenshots/animated gifs/videos if necessary._

@stacy's fix:

https://user-images.githubusercontent.com/1581694/135698513-834435c9-3ad3-4756-9fc3-1fb14615829e.mov

dvdherron commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Top line of Scratch Org Cards run out of room (mostly when two buttons are shown) at certain screen sizes](https://trello.com/c/tLZcyRMS/324-top-line-of-scratch-org-cards-run-out-of-room-mostly-when-two-buttons-are-shown-at-certain-screen-sizes)

dvdherron commented:
@stacyk Still working through how to fix this. If you have time later today could we chat about it?

stacyk commented:
@dvdherron: you are right, container queries would be so helpful for this one... What I am proposing is not perfect, but without re-writing the card header in grid, and without being able to edit too much of the markup, this is what I came up with: 

- I added a narrow-buttons class to edit button padding at medium + screen widths
- I am allowing the buttons container to wrap
- I added a minimum width for the card/org name so it doesn't truncate down to nothing

The thing I like least is the dropdown trigger wraps, and then is a little awkward at in between sizes, but I think a decent trade off for buttons to not collide. 

https://user-images.githubusercontent.com/1581694/135521785-0dc929d1-58cb-4921-885d-0af212e33c88.mov

dvdherron commented:
> 
> 
> @dvdherron: you are right, container queries would be so helpful for this one... What I am proposing is not perfect, but without re-writing the card header in grid, and without being able to edit too much of the markup, this is what I came up with:
> 
>     * I added a narrow-buttons class to edit button padding at medium + screen widths
> 
>     * I am allowing the buttons container to wrap
> 
>     * I added a minimum width for the card/org name so it doesn't truncate down to nothing
> 
> 
> The thing I like least is the dropdown trigger wraps, and then is a little awkward at in between sizes, but I think a decent trade off for buttons to not collide.
> metecho-button-layout-proposal.mov

This looks great @stacyk! Trying to see if there's something to do about the dropdown wrapping on the project detail page. For me, the dropdown doesn't wrap like that on an epic/task detail page.

stacyk commented:
> Is it better or worse if that that dropdown button doesn't have any left margin when space runs out?

I don't know. I originally wanted to have that button not move at all. I just pushed an update that absolutely positions the dropdown trigger in the top right. Want to take a look and see if you like it? I will upload a video later tonight.

stacyk commented:
Here's what I see in Chrome with the absolutely positioned button. I know the 1-2 brand buttons move together, but I think that is ok? What are your thoughts @dvdherron:

https://user-images.githubusercontent.com/1581694/135698513-834435c9-3ad3-4756-9fc3-1fb14615829e.mov

dvdherron commented:
> 
> 
> Here's what I see in Chrome with the absolutely positioned button. I know the 1-2 brand buttons move together, but I think that is ok? What are your thoughts @dvdherron:
> metecho-buttons-part2.mov

I think this works great @stacyk. It makes a lot more sense for the dropdown button to stay at the top of the card no matter where the "contribute work" and "refresh org" buttons end up when they wrap. I think this solves this issue?

mirisuzanne commented on src/sass/components/_org-cards.scss:
```diff
@@ -13,20 +13,38 @@
+      display: flex;
+      flex: 0 1 auto;
+      flex-wrap: wrap; // allow button group to wrap when space is limited
+      margin-right: $spacing-medium;
```
I assume this is leaving space for the `metecho-card-more` dropdown? Maybe worth commenting to clarify that the sizes are related?

stacyk replied:
@mirisuzanne I made a few changes since your comment as it made me think about how to simplify and keep the margin consistent. Instead, I edited the padding end variable to give the button enough space on the right instead of larger right margins on the title container and button containers (they both now have a small margin which looks better anyway) :)

mirisuzanne reviewed (approved):
I like the final result here. Nice work.

jgerigmeyer reviewed (changes requested):
The changes in `orgActions` apply to all orgs (dev, test, scratch). The changes in `playgroundCard` only apply to scratch orgs. Some of the changes here are breaking the actions layout for dev/test orgs:

![Screen Shot 2021-10-04 at 6 36 06 PM](https://user-images.githubusercontent.com/552316/135933962-5fb2a503-2205-430c-ba8c-a52eee5b0a7f.png)

dvdherron commented:
@stacyk 

The changes I've made keep the buttons on a separate row for a slightly longer time but I think this layout works for all cases. Can you take a look?

![image](https://user-images.githubusercontent.com/41588129/136053791-9a0f4f26-939a-49fd-8837-2e1de8e5dc8f.png)

![image](https://user-images.githubusercontent.com/41588129/136053711-1e938110-19cb-487d-8e33-5f816e818f59.png)

jgerigmeyer commented:
> The changes I've made keep the buttons on a separate row for a slightly longer time but I think this layout works for all cases.

@dvdherron I don't think Dev/Test orgs actually get a "Contribute Work" btn. It's not clear, so let me try to clarify the various possible btns:

Scratch Orgs:

- "Creating Org…" or "Refreshing Org…" (if either of these is shown, no other btns will be shown)
- "Refresh Org"
- "Contribute Work"

Dev Orgs:

- "Creating Org…" (if this is shown, no other btns will be shown)

Test Orgs:

- "Creating Org…" or "Refreshing Org…" or "Submitting Review…" (if one of these is shown, no other btns will be shown)
- "Refresh Org"
- "Update Review" or "Submit Review"

stacyk commented on src/js/components/orgs/cards/orgActions.tsx:
```diff
@@ -131,7 +134,7 @@ const OrgActions = ({
             label={i18n.t('Refresh Org')}
             variant="brand"
-            className="slds-m-horizontal_x-small"
+            className="slds-m-right_x-small"
```
I think this one could benefit from `slds-m-left_none`

jgerigmeyer commented on src/sass/components/_org-cards.scss:
```diff
@@ -1,4 +1,6 @@
 .has-nested-card {
+  // remove extra space on org action dropdown
+  --sds-c-card-body-spacing-inline-end: 0;
```
This makes me nervous... isn't it an undocumented part of SLDS that could change?

stacyk replied:
Darn, yes, it is in "beta" at the moment and I didn't catch that. I would be surprised if they don't go that way but we can update this here, and there is another place I am using them (in buttons.scss I think).

jgerigmeyer commented on src/js/components/orgs/cards/orgActions.tsx:
```diff
@@ -118,7 +118,10 @@ const OrgActions = ({
-        className="slds-m-right_x-small tour-scratch-org-contribute"
+        className="slds-m-right_x-small
+          slds-m-left_none
+          slds-m-top_xx-small
```
I don't think this top-margin gains us much, and it makes the btn and the dropdown no longer aligned.

![Screen Shot 2021-10-06 at 3 45 52 PM](https://user-images.githubusercontent.com/552316/136272741-32abd468-7c66-49b4-9dce-98f4b96bf637.png)

jgerigmeyer replied:
Yeah, that was my assumption -- but even when stacked, I wasn't sure it was necessary. 🤷

stacyk replied:
yeah, we have row-gap too so I can remove it. I think I was going back and forth there for a while.

stacyk reviewed (commented):
@dvdherron I have not been able to get to this yet so feel free to continue on it if you want something to work on.

mirisuzanne commented on src/sass/components/_org-cards.scss:
```diff
@@ -10,23 +10,51 @@
+    // Specifically target OrgActions in card header
+    .slds-no-flex {
+      align-items: baseline;
+      display: flex;
```
Adding `display: flex` to `.slds-no-flex` feels pretty weird to me. Is there any way we can add a new class for this override - or remove that one?

stacyk replied:
@mirisuzanne It feels gross to me too, and I know we don't like to change the meaning of one of their classes, but we are at least nesting it in our own `.wrap-inner-truncate` class if that helps a tiny bit. @jgerigmeyer correct me if I am wrong on this, but I don't think we have access to the element with the `slds-no-flex` class.

jgerigmeyer replied:
> @jgerigmeyer correct me if I am wrong on this, but I don't think we have access to the element with the `slds-no-flex` class.

That's correct.

mirisuzanne replied:
ok, yeah, probably nothing better do do about it then…

mirisuzanne reviewed (commented):
Generally looks good - just a bit worried about `slds-no-flex` doing the opposite of what it says.
